docs: clarify requirement to explicitly specify storage class for vault on rke2 (PAC-829) (#1733)
spectrocloud/librarium: docs/docs-content/integrations/vault.md

Reviewer: Terraform section is needed.

## Versions Supported

<Tabs queryString="versions">

<TabItem label="0.22.x" value="0.22.x">

### Prerequisites

- A Kubernetes cluster with Kubernetes version 1.22 or later.

### Parameters

The table lists commonly used parameters you can configure when adding this pack.

| Parameters | Description | Default |
-------------|-------------|---------|
|`charts.server.ingress` | Enable ingress traffic to the Vault server. If you want to enable ingress traffic, make sure that `charts.server.serviceType` is set to `"ClusterIP"` or is left empty. | `false` |
|`charts.global.tlsDisable` | Disable TLS for end-to-end encrypted transport. | `true` |
|`charts.agent.enabled` | Enable vault agent injection to inject secrets into the pods. | `-` |
|`charts.server.ha` | Enable high-availability mode to protect against outages by running multiple Vault servers. For more information, refer to [Vault documentation](https://developer.hashicorp.com/vault/docs/internals/high-availability). |  `false` |
|`charts.server.dataStorage`| Controls the size, location, storage class of the persistent storage used by the Vault. | |

> :::caution

Reviewer 2: This caution is very specific to RKE2. Could we create a section in usage for maybe Storage and mention this under that section? I get the sense we might have more of this challenges related to storage in the future.

You can do so by following these steps:

1. Log in to [Palette](https://console.spectrocloud.com).

2. Navigate to the left **Main Menu** and select **Clusters**.

3. Select the cluster that has Vault installed to view its details page.

4. Download the cluster **kubeconfig** file. 

5. Set up your local kubectl environment to use the **kubeconfig** file you downloaded. Review the [Access Cluster with CLI](../clusters/cluster-management/palette-webctl.md) guide for additional guidance. 

6. You need to get the Vault namespace and application name. Issue the following command to get the unique values.

    <br />

    ```shell
    VAULT_NAMESPACE=$(kubectl get pods --selector app.kubernetes.io/name=vault --all-namespaces --output jsonpath='{.items[0].metadata.namespace}') && \
    APP_NAME=$(echo "$VAULT_NAMESPACE" | sed 's/-ns$//')
    ```

7. Set up port forwarding by issuing the following command so you can access the Vault UI:

    ```
    kubectl port-forward $APP_NAME 8200:8200 --namespace $VAULT_NAMESPACE
    ```

> 8. Open your browser and access the [Vault UI](https://localhost:8200/ui. You will receive a warning due to using a self-signed certificate, but you can ignore this warning. Follow the prompts on the UI to initialize your root token.

Reviewer 2: ```suggestion
8. Open your browser and access the [Vault UI](https://localhost:8200/ui). You will receive a warning due to using a self-signed certificate, but you can ignore this warning. Follow the prompts on the UI to initialize your root token.
```
added missing parentheses.

> 8. Open your browser and access the [Vault UI](https://localhost:8200/ui. You will receive a warning due to using a self-signed certificate, but you can ignore this warning. Follow the prompts on the UI to initialize your root token.

Reviewer 2: @lennessyy What is the root token? Can I set the root token value through the YAML? Can I get it by reviewing a kubernetes secret?

Author: I can add a brief definition of root tokens, but I don't think you can set root token value through the YAML unless you are using Dev mode instead of Prod mode. If you enable Dev mode, then yes, you can set it in the YAML and can probably set some other configuration to get it by reviewing Kubernetes secret, but otherwise I think you have to go through initialization:
https://developer.hashicorp.com/vault/docs/concepts/tokens#root-tokens

Author: I can add dev mode to the list of parameters so people know

:::tip

> If you don't want to use the Vault UI, you can also initialize and unseal Vault using the Vault CLI or API. For more information, refer to [Vault documentation](https://developer.hashicorp.com/vault/docs/platform/k8s/helm/run#initialize-and-unseal-vault).  

Reviewer: ```suggestion
If you do not want to use the Vault UI, you can also initialize and unseal Vault using the Vault CLI or API. For more information, refer to [Vault documentation](https://developer.hashicorp.com/vault/docs/platform/k8s/helm/run#initialize-and-unseal-vault).  
```

Reviewer: We don't typically use contractions in the docs.

:::

### Terraform

You can reference the Vault pack in Terraform with a data resource:

> ```tf

Reviewer: ```suggestion
```hcl
```

Reviewer: The Terraform example doesn't display like the others in the docs. Maybe if you use `hcl`? I looked at the kubernetes example for PXK.

6. You need to get the Vault namespace and application name. Issue the following command to get the unique values.

    <br />

    ```shell
    VAULT_NAMESPACE=$(kubectl get pods --selector app.kubernetes.io/name=vault --all-namespaces --output jsonpath='{.items[0].metadata.namespace}') && \
    APP_NAME=$(echo "$VAULT_NAMESPACE" | sed 's/-ns$//')
    ```

7. Set up port forwarding by issuing the following command so you can access the Vault UI:

    ```
    kubectl port-forward $APP_NAME 8200:8200 --namespace $VAULT_NAMESPACE
    ```

8. Open your browser and access the [Vault UI](https://localhost:8200/ui). You will receive a warning due to using a self-signed certificate, but you can ignore this warning. Follow the prompts on the UI to initialize your root token.

:::tip

If you do not want to use the Vault UI, you can also initialize and unseal Vault using the Vault CLI or API. For more information, refer to [Vault documentation](https://developer.hashicorp.com/vault/docs/platform/k8s/helm/run#initialize-and-unseal-vault).  

:::

#### Storage

> In a production Vault server, backend storage is on a data persistent layer, is untrusted and only stores encrypted data. In a dev mode Vault server, all data is stored in-memory and will be erased when Vault restarts. 

Reviewer: ```suggestion
In a production Vault server, backend storage is on a data persistent layer, is untrusted, and only stores encrypted data. In a dev mode Vault server, all data is stored in-memory and will be erased when Vault restarts. 
```

* UI (Optional).
* [Agent injector](https://www.vaultproject.io/docs/platform/k8s/injector/) (Optional).

:::caution

When using Vault with the RKE2 distribution of Kubernetes in Palette Edge, you must explicitly specify a storage class for the Vault server. To specify a storage class, insert the following fields under the `server` field in `values.yaml` for the Vault layer in your cluster profile. 

```yaml
server:
    dataStorage:

>         storageClass: <storage class name>

Author: @vishwanaths Can you confirm this is the right place to add the storage class? In addition, it would be helpful to let users know what happens if they don't specify it.
